File: code/HuaWeiNote/app/src/main/java/com/fan/hwnote/app/model/NotebookRepository.kt
package com.fan.hwnote.app.model

import android.content.ContentValues
import android.content.Context
import com.fan.hwnote.app.model.db.NoteDbHelper
import com.fan.hwnote.app.model.entity.Notebook
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object NotebookRepository {

    private lateinit var helper: NoteDbHelper

    fun init(context: Context) {
        helper = NoteDbHelper(context.applicationContext)
    }

    suspend fun listByFolder(folderId: Long): List<Notebook> = withContext(Dispatchers.IO) {
        val out = mutableListOf<Notebook>()
        helper.readableDatabase.rawQuery(
            "SELECT id, folder_id, name, color, order_index, is_default, deleted_at FROM notebooks WHERE folder_id = ? AND deleted_at = 0 ORDER BY is_default DESC, order_index ASC, id ASC",
            arrayOf(folderId.toString())
        ).use { c ->
            while (c.moveToNext()) out += cursorToNotebook(c)
        }
        out
    }

    suspend fun get(id: Long): Notebook? = withContext(Dispatchers.IO) {
        helper.readableDatabase.rawQuery(
            "SELECT id, folder_id, name, color, order_index, is_default, deleted_at FROM notebooks WHERE id = ? AND deleted_at = 0",
            arrayOf(id.toString())
        ).use { c -> if (c.moveToFirst()) cursorToNotebook(c) else null }
    }

    suspend fun insert(folderId: Long, name: String, color: String): Long = withContext(Dispatchers.IO) {
        val db = helper.writableDatabase
        val nextIndex = nextOrderIndex(db, folderId)
        val cv = ContentValues().apply {
            put("folder_id", folderId)
            put("name", name)
            put("color", color)
            put("order_index", nextIndex)
            put("is_default", 0)
            put("deleted_at", 0L)
        }
        db.insert("notebooks", null, cv)
    }

    suspend fun rename(id: Long, name: String) = withContext(Dispatchers.IO) {
        val cv = ContentValues().apply { put("name", name) }
        helper.writableDatabase.update("notebooks", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    suspend fun updateColor(id: Long, color: String) = withContext(Dispatchers.IO) {
        val cv = ContentValues().apply { put("color", color) }
        helper.writableDatabase.update("notebooks", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    suspend fun softDelete(id: Long) = withContext(Dispatchers.IO) {
        require(id != 1L) { "默认笔记本不可删除" }
        val now = System.currentTimeMillis()
        val db = helper.writableDatabase
        db.beginTransaction()
        try {
            db.execSQL(
                "UPDATE notes SET deleted_at = ? WHERE deleted_at = 0 AND notebook_id = ?",
                arrayOf<Any>(now, id)
            )
            val cv = ContentValues().apply { put("deleted_at", now) }
            db.update("notebooks", cv, "id = ?", arrayOf(id.toString()))
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
        Unit
    }

    suspend fun move(id: Long, targetFolderId: Long) = withContext(Dispatchers.IO) {
        require(id != 1L) { "默认笔记本不可跨文件夹移动" }
        val db = helper.writableDatabase
        val nextIndex = nextOrderIndex(db, targetFolderId)
        val cv = ContentValues().apply {
            put("folder_id", targetFolderId)
            put("order_index", nextIndex)
        }
        db.update("notebooks", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    suspend fun reorderInFolder(folderId: Long, orderedIds: List<Long>) = withContext(Dispatchers.IO) {
        val db = helper.writableDatabase
        db.beginTransaction()
        try {
            orderedIds.forEachIndexed { idx, id ->
                if (id == 1L && folderId == 1L) return@forEachIndexed
                val cv = ContentValues().apply { put("order_index", idx) }
                db.update("notebooks", cv, "id = ? AND folder_id = ?", arrayOf(id.toString(), folderId.toString()))
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
        Unit
    }

    private fun nextOrderIndex(db: android.database.sqlite.SQLiteDatabase, folderId: Long): Int {
        db.rawQuery(
            "SELECT IFNULL(MAX(order_index), -1) + 1 FROM notebooks WHERE deleted_at = 0 AND folder_id = ?",
            arrayOf(folderId.toString())
        ).use { c -> return if (c.moveToFirst()) c.getInt(0) else 0 }
    }

    private fun cursorToNotebook(c: android.database.Cursor): Notebook = Notebook(
        id = c.getLong(0),
        folderId = c.getLong(1),
        name = c.getString(2),
        color = c.getString(3),
        orderIndex = c.getInt(4),
        isDefault = c.getInt(5) == 1,
        deletedAt = c.getLong(6),
    )
}

File: code/HuaweiNoteFlutter/lib/repositories/notebook_repository.dart
import 'package:sqflite/sqflite.dart';
import '../db/database_helper.dart';
import '../models/notebook.dart';

class NotebookRepository {
  final DatabaseHelper _dbHelper;

  NotebookRepository(this._dbHelper);

  Future<List<Notebook>> listByFolder(int folderId) async {
    final db = await _dbHelper.database;
    final rows = await db.rawQuery(
      'SELECT id, folder_id, name, color, order_index, is_default, deleted_at FROM notebooks WHERE folder_id = ? AND deleted_at = 0 ORDER BY is_default DESC, order_index ASC, id ASC',
      [folderId],
    );
    return rows.map(_rowToNotebook).toList();
  }

  Future<Notebook?> get(int id) async {
    final db = await _dbHelper.database;
    final rows = await db.rawQuery(
      'SELECT id, folder_id, name, color, order_index, is_default, deleted_at FROM notebooks WHERE id = ? AND deleted_at = 0',
      [id],
    );
    if (rows.isEmpty) return null;
    return _rowToNotebook(rows.first);
  }

  Future<int> insert(int folderId, String name, String color) async {
    final db = await _dbHelper.database;
    final nextIndex = await _nextOrderIndex(db, folderId);
    return db.insert('notebooks', {
      'folder_id': folderId, 'name': name, 'color': color,
      'order_index': nextIndex, 'is_default': 0, 'deleted_at': 0,
    });
  }

  Future<void> rename(int id, String name) async {
    final db = await _dbHelper.database;
    await db.update('notebooks', {'name': name}, where: 'id = ?', whereArgs: [id]);
  }

  Future<void> updateColor(int id, String color) async {
    final db = await _dbHelper.database;
    await db.update('notebooks', {'color': color}, where: 'id = ?', whereArgs: [id]);
  }

  Future<void> softDelete(int id) async {
    assert(id != 1, '默认笔记本不可删除');
    final db = await _dbHelper.database;
    final now = DateTime.now().millisecondsSinceEpoch;
    await db.transaction((txn) async {
      await txn.rawUpdate(
        'UPDATE notes SET deleted_at = ? WHERE deleted_at = 0 AND notebook_id = ?',
        [now, id],
      );
      await txn.update('notebooks', {'deleted_at': now}, where: 'id = ?', whereArgs: [id]);
    });
  }

  Future<void> move(int id, int targetFolderId) async {
    assert(id != 1, '默认笔记本不可跨文件夹移动');
    final db = await _dbHelper.database;
    final nextIndex = await _nextOrderIndex(db, targetFolderId);
    await db.update('notebooks', {
      'folder_id': targetFolderId, 'order_index': nextIndex,
    }, where: 'id = ?', whereArgs: [id]);
  }

  Future<void> reorderInFolder(int folderId, List<int> orderedIds) async {
    final db = await _dbHelper.database;
    await db.transaction((txn) async {
      for (var i = 0; i < orderedIds.length; i++) {
        final id = orderedIds[i];
        if (id == 1 && folderId == 1) continue;
        await txn.update('notebooks', {'order_index': i},
          where: 'id = ? AND folder_id = ?', whereArgs: [id, folderId]);
      }
    });
  }

  Future<int> _nextOrderIndex(Database db, int folderId) async {
    final result = await db.rawQuery(
      'SELECT IFNULL(MAX(order_index), -1) + 1 FROM notebooks WHERE deleted_at = 0 AND folder_id = ?',
      [folderId],
    );
    return Sqflite.firstIntValue(result) ?? 0;
  }

  Notebook _rowToNotebook(Map<String, Object?> row) => Notebook(
    id: row['id'] as int,
    folderId: row['folder_id'] as int,
    name: row['name'] as String,
    color: row['color'] as String,
    orderIndex: row['order_index'] as int,
    isDefault: (row['is_default'] as int) == 1,
    deletedAt: row['deleted_at'] as int? ?? 0,
  );
}
